rust/main/hyperlane-core/src/chain.rs

```diff
@@ -39,6 +42,80 @@ impl<'a> std::fmt::Display for ContractLocator<'a> {
     }
 }
 
+#[derive(Default, Debug, Clone, PartialEq)]
+pub enum ReorgPeriod {
+    #[default]
+    None,
+    Blocks(NonZeroU32),
+    Tag(String),
+}
+
+impl ReorgPeriod {
+    pub fn from_number(number: u32) -> Self {
+        NonZeroU32::try_from(number)
+            .map(ReorgPeriod::Blocks)
+            .unwrap_or(ReorgPeriod::None)
+    }
+
+    pub fn as_number(&self) -> Result<u32, Report> {
+        match self {
+            ReorgPeriod::None => Ok(0),
+            ReorgPeriod::Blocks(blocks) => Ok(blocks.get()),
+            ReorgPeriod::Tag(_) => Err(eyre!("Invalid reorg period")),
+        }
+    }
+
+    pub fn is_none(&self) -> bool {
+        matches!(self, ReorgPeriod::None)
+    }
+}
+
+impl Serialize for ReorgPeriod {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
```

Comment:
IMO, `ReorgPeriod::Tag` should be defined as an enum. Also the serialization should look like the one below, which I believe you can configure `derive(Serialize)` to give you with the flattening flag

```
            ReorgPeriod::None => "None"
            ReorgPeriod::Blocks(blocks) => blocks.to_string(),
            ReorgPeriod::Tag(tag) => tag.to_string()
```

Then to deserialize first you try to find the string in the predefined `Tag` enum, then match it against the `None` string, or if it's a number it's parsed as blocks (probably harder to define via the `derive`)

Reply:
@daniel-savu it was @tkporter's [idea to have it as a string](https://discord.com/channels/935678348330434570/1285726511793704971/1288917019550289941):

> so I'd prefer
> 
> enum ReorgConfiguration {
>   Blocks(u64),
>   Tag(String),
> }
> 
> instead of
> 
> enum ReorgConfiguration {
>   Blocks(u64),
>   Tag(TagType),
> }
> 
> enum TagType {
>   Finalized,
>   Safe
> }

Reply:
Yeah as this type is meant to be protocol agnostic I think having it be as flexible as possible is in our favor. Otherwise I guess we'd need this to be a union of all protocol supported tags, which feels a bit awk

My original motivation though was hoping that ethers-rs would let us pass in arbitrary tags because I've seen some chains use new tags outside of the typical ones, but sadly it looks like ethers-rs has their own enum they force you to use. But maybe on other chains this won't always be the case

Reply:
you probably need to support conversions for each individual tag anyway (as happens now in `TryFrom<&ReorgPeriod> for EthereumReorgPeriod`), so you need to deal with all tags variants either way, it's just that the parsing happens further into execution. 

I'm ok with allowing arbitrary strings, but mostly because the `TryFrom`s live in chain-specific integration crates, rather than having an enum that uses types from all these chain SDKs in `hyperlane-core`/`hyperlane-base`